### PR TITLE
php: add config.php.mysqlnd option

### PR DESCRIPTION
**Pull request by jbboehr**
###### Motivation for this change

Adds support for PHP's native MySQL driver. See #31167 - issues using mysqlclient. This PR is meant as an example for my proposal.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I've only tested that the extensions loaded and displayed the proper driver version, not it's actual functionality, nor any of the other PHP functionality.

### CODE REVIEW

**Comment by jbboehr** [comment]
Looks like the Travis build failed due to #20491

**Comment by jbboehr** [comment]
Got the error: `configure: error: Cannot find MySQL header files under . Note that the MySQL client library is not bundled anymore!`

**Comment by jbboehr** [comment]
Error was due to a misplaced `=`

**Comment by jbboehr** [comment]
Well, Travis seems to be failing and I don't think it's my fault (famous last words), so I will try merging master in later.

**Comment by jbboehr** [comment]
Travis should be passing now.

@orivej @Mic92 Any thoughts on this PR?

File: pkgs/development/interpreters/php/default.nix [hunk 1]
```diff
@@ -11,6 +11,8 @@ let
 
     let php7 = lib.versionAtLeast version "7.0";
         mysqlHeaders = mysql.lib.dev or mysql;
+        mysqlndSupport = config.php.mysqlnd or false;
+        mysqlBuildInputs = if mysqlndSupport then [] else [ mysqlHeaders ];
```
Comment (Mic92) [inline pkgs/development/interpreters/php/default.nix]: -        mysqlBuildInputs = if mysqlndSupport then [] else [ mysqlHeaders ]; 
+       mysqlBuildInputs = lib.optional mysqlndSupport mysqlHeaders;

Comment (orivej) [reply pkgs/development/interpreters/php/default.nix]: You've reversed the conditional.

**Comment by jbboehr** [comment]
> A downside of your approach is that config.php.mysqld is set aside from other configuration options and is not as easily discoverable

To be completely honest, I have no idea how the `cfg` section or `composableDerivation` work. I would've added it to the `cfg` section but the enable switches need to be passed the value `mysqlnd`, and there was a precedent for passing it in in the location I chose in this PR.

> I have not seen composableDerivation machinery before, and I was not able to achieve what you are trying to do here with it.

Can you clarify what you mean by 'I was not able to achieve what you are trying to do here with it'?

Thanks for taking the time to review this PR, @orivej @Mic92

**Review by orivej — APPROVED** [review]
> Can you clarify what you mean by 'I was not able to achieve what you are trying to do here with it'?

I've read the definition of `composableDerivation` and tried to move `mysqlnd` into `cfg`. In the end this failed because I could not read `cfg` from flags and I could not affect one flag from another.

**Comment by jbboehr** [comment]
@orivej Ah ok. I believe I tried the same thing originally and experienced the same issue.

File: pkgs/development/interpreters/php/default.nix (hunk 1 above)
Comment (jbboehr) [reply pkgs/development/interpreters/php/default.nix]: Applied in 1cff740

**Pull request merged**